samples-lib/src/main/java/com/esri/arcgismaps/sample/sampleslib/DownloaderActivity.kt

```diff
@@ -224,43 +233,38 @@ abstract class DownloaderActivity : AppCompatActivity() {
                 emit(LoadStatus.FailedToLoad(it))
                 return@flow
             }
+            // set up the file at the download path
+            val destinationFilePath = provisionLocation.path + File.separator + portalItem.name
+            // set up the download URL
+            val downloadURL =
+                "${portalItem.portal.url}/sharing/rest/content/items/${portalItem.itemId}/data"
             // get the data of the PortalItem
-            val portalItemData = portalItem.fetchData()
-            val byteArray = portalItemData.getOrElse {
-                // close dialog and emit status
-                loadingDialog.dismiss()
-                emit(LoadStatus.FailedToLoad(it))
-                return@flow
-            }
-            // get the byteArray of the PortalItem
-            runCatching {
-                val byteArrayInputStream = ByteArrayInputStream(byteArray)
-                val data = ByteArray(1024)
-                var downloadCount: Int
-                downloadCount = byteArrayInputStream.read(data)
-                while (downloadCount != -1) {
-                    downloadCount = byteArrayInputStream.read(data)
+            ArcGISEnvironment.arcGISHttpClient.download(
+                url = downloadURL,
+                destinationFile = File(destinationFilePath)
+            ) { totalBytes, bytesRead ->
+                if (totalBytes != null) {
+                    // update the download percentage progress
+                    val percentage = ((100.0 * bytesRead) / totalBytes).roundToInt()
+                    Handler(Looper.getMainLooper()).post {
```

Comment:
Interesting. This should not be necessary. I wonder if we should make the progress listener lambda a suspend function, in which case you could just call something like:
```kotlin
withContext(Dispatchers.Main) {
   progressIndicator.progress = percentage
   progressTV.text = "$percentage%"
}
```

Alternativey we could make sure that the progress listener is always invoked on the main dispatcher.

@changanxian - thoughts?

Reply:
Updated this to use your suggestion. This is required since changes to the linear progress indicator is required to run on the main thread:
```kotlin
lifecycleScope.launch(Dispatchers.Main) {
    progressIndicator.progress = percentage
    progressTV.text = "$percentage%"
}

Reply:
Since the listener is updating the progress UI, it needs to run on the `main` thread.  Changing the listener be `suspend` function makes more sense to me.
@gunt0001

Reply:
In my mind, invoking the progress listener on the main dispatcher would be evn more convenient, as the user would always be called back on the main dispatcher, so non need at all for any thread switching in user code.